finality-aleph/src/testing/mocks/validator_network.rs

```diff
@@ -90,3 +112,373 @@ impl<D: Data> MockNetwork<D> {
+        MockData {
+            data,
+            filler: vec![0; filler_size],
+            decodes,
+        }
+    }
+}
+
+impl Encode for MockData {
+    fn encode(&self) -> Vec<u8> {
+        // currently this is exactly the default behaviour of codec
+        let mut encoded = Vec::with_capacity(self.filler.len() + 5);
+        encoded.append(&mut u32::encode(&self.data));
+        encoded.append(&mut Vec::<u8>::encode(&self.filler));
+        encoded.append(&mut bool::encode(&self.decodes));
+        encoded
+    }
+}
+
+impl Decode for MockData {
+    fn decode<I: codec::Input>(value: &mut I) -> Result<Self, codec::Error> {
+        let data = u32::decode(value)?;
+        let filler = Vec::<u8>::decode(value)?;
+        let decodes = bool::decode(value)?;
+        if !decodes {
+            return Err("Simulated decode failure.".into());
+        }
+        Ok(Self {
+            data,
+            filler,
+            decodes,
+        })
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn spawn_peer(
+    pen: AuthorityPen,
+    addr: Addresses,
+    n_msg: usize,
+    large_message_interval: usize,
+    corrupted_message_interval: usize,
+    dialer: MockDialer,
+    listener: MockListener,
+    report: mpsc::UnboundedSender<(AuthorityId, usize)>,
+    spawn_handle: SpawnTaskHandle,
+) {
+    let our_id = pen.authority_id();
+    let (service, mut interface) = Service::new(dialer, listener, pen, spawn_handle);
+    // run the service
+    tokio::spawn(async {
+        let (_exit, rx) = oneshot::channel();
+        service.run(rx).await;
+    });
+    // start connecting with the peers
+    let mut peer_ids = Vec::with_capacity(addr.len());
+    for (id, addrs) in addr.into_iter() {
+        interface.add_connection(id.clone(), addrs);
+        peer_ids.push(id);
+    }
+    // peer main loop
+    // we send random messages to random peers
+    // a message is a number in range 0..n_msg
+    // we also keep a list of messages received at least once
+    // on receiving a message we report the total number of distinct messages received so far
+    // the goal is to receive every message at least once
+    tokio::spawn(async move {
+        let mut received: Vec<bool> = vec![false; n_msg];
+        let mut send_ticker = tokio::time::interval(Duration::from_millis(5));
+        let mut counter: usize = 0;
+        loop {
+            tokio::select! {
+                _ = send_ticker.tick() => {
+                    counter += 1;
+                    // generate random message
+                    let filler_size = if counter % large_message_interval == 0 { 32*1024*1024 } else { 0 };
+                    let decodes = counter % corrupted_message_interval != 0;
+                    let data: MockData = MockData::new(thread_rng().gen_range(0..n_msg) as u32, filler_size, decodes);
+                    // choose a peer
+                    let peer: AuthorityId = peer_ids[thread_rng().gen_range(0..peer_ids.len())].clone();
```

Comment:
This looks like a coupon collector problem. I would add somewhere a note about this taking on average `n log(n)` messages to collect everything. Where `n` is `n_msg * n_peers`.

Reply:
added docs to the `scenario` method

Reply:
I thought n log(n) comment would be more appropriate here, as this is the  confusing part of code, but I guess dosc to the scenario method are fine too.

Reply:
yeah, I think this is an important remark, so it should be more visible